### ISSUE TITLE
Iteration is a nuisance

### ISSUE THREAD

**Issue by vanrein** [vanrein/quick-der#74]
The C interface for iteration is mind-boggling.  Even when it is sort-of okay to write...

```
if (der_iter_first (&crs, &iter) do {
   ...work on iter...
} while (der_iter_next (&iter));
```

...there still is confusion about whether to `der_enter()` and `der_focus()`, both inside and prior to the iteration.  The cause is basically that the iteration uses `.derlen` that span to the end of iteration, which is not normally what we use.  Or one might say that the `.derptr` has dived in one more level than `.derlen`.

A normal API for iteration would use a callback, and that seems quite doable here too; it requires more functions but simplifies the conceptual use of iteration.  We might also improve on the current API, of course, for instance by taking the `.derlen` from the original cursor and thus require it to be presented in `der_iter_next()` too.

**Comment by vanrein** [comment]
There was a mixed-up concept in the code!

`der_header()` enters into the analysed header.  Note sure why, but that's what it has been doing all along.  This is not always desirable, so I created `der_header2()` that works on a copied argument, so `dercursor` instead of `dercursor *` and that can remedy this.

The wrong interpretation was woven into `der_focus()` and `der_skip()`, which in turn impacted `der_iterate_first()` and `der_iterate_next()`.  Since these were plain wrong, I fixed them.

**This creates an incompatibility** but a quick scan of the code using iterators revealed that this is precisely the problem that has been worked around.  The code usually shows the intention, comments it out and works around it.  This is only used in a few places anyway, so the incompatible break is basically a resolution of a bug, therefore let through.  *Be aware that iteration code now needs some fixup towards sanity!*
